texts/0004-mobile-plugin.md

```diff
@@ -0,0 +1,201 @@
+  precision: usize,
+}
+
+#[derive(serde::Deserialize)]
+struct CurrentLocation {
+  location: String,
+}
+
+let payload = GetCurrentLocationPayload {
+  precision: 1,
+};
+// Run the plugin command `getCurrentLocation` with the given payload (any Serialize type)
+let response: CurrentLocation = handle.run_mobile_plugin("getCurrentLocation", payload).unwrap();
+```
+
+The default plugin template will include this boilerplate, storing the plugin handle in the Tauri state using the newtype idiom so each plugin has its own handle.
+
+Additionally, the mobile plugin command can be reached directly by the webview side:
+
+```javascript
+import { invoke } from '@tauri-apps/api/invoke'
+const { location } = await invoke('plugin:example|getCurrentLocation', { precision: 0 })
+```
+  
+# Reference-level explanation
+
+To achieve this functionality, we need to change the existing codebase:
+
+## Plugin Template
+
+- Adapted to support both mobile and desktop
+- iOS and Android projects
+
+## CLI
+
+- New commands to add Android and iOS projects to an existing plugin
+- On dev and build, inject the `Tauri` dependency on each plugin so they can be compiled
+  - This dependency exposes the native plugin API for Android and iOS and is injected dynamically so it's always in sync
+
+## tauri
+
+- New plugin APIs for registering and calling mobile plugin commands `PluginApi` and `PluginHandle`
+- The IPC handler now sends a message to the native Kotlin and iOS IPC handlers if Rust did not respond to the command
+  - With this change, the JavaScript layer can directly reach the native mobile APIs
+- New macro to generate the function that initializes the iOS plugin `ios_plugin_binding!(fn_name)`
+- Plugin payloads are converted to `serde_json::Value` and a Rust code converts that to a Swift dictionary using Objective-C or a Java map using JNI.
+- Plugin responses are serialized by the native code and deserialized on Rust :(
+
+## tauri-build
+
+- New `mobile` module with a builder to run the Android and iOS build pipelines for a plugin
+  - For Android, injects the Android Studio project to the application folder
+  - On iOS, build and link the Swift package
+
+## Platform Considerations
+
+Plugin and user code must be careful with the target platforms when writing code for mobile. `#[cfg(target_os = "ios")]` and `#[cfg(target_os = "android")]` will be widely used, and the default plugin template takes care of the initial boilerplate for this, also preparing the plugin code to work on desktop by introducing a separate module to write the same functionality in Rust so macOS, Linux and Windows applications can call the same APIs. So to support all platforms, users need to write both Java, Swift and Rust code, taking care of using each platform's APIs to reach the plugins goals. This introduces a lot of work for plugin maintainers, but has a lot of benefits for application developers that can use all the functionality with an unified API.
```

Comment:
We can use [cfg_aliases](https://docs.rs/cfg_aliases/latest/cfg_aliases/) to rename a few target os. Maybe we should use it in all repo since we need to deal with different ports a lot.

Reply:
We have a feature flag alias in tauri-build:

```rust
fn alias(alias: &str, has_feature: bool) {
  if has_feature {
    println!("cargo:rustc-cfg={alias}");
  }
}
```

Currently we have a `desktop` and a `mobile` alias.